### ISSUE TITLE
Please make high resolution time available within AudioWorkletGlobalScope.

### ISSUE THREAD

**Issue by issuefiler** [WebAudio/web-audio-api#2413]
# The desired feature

 [High resolution time](https://www.w3.org/TR/hr-time-2) available within `AudioWorkletGlobalScope`.

> This specification defines an API that provides the time origin, and current time in sub-millisecond resolution, such that it is not subject to system clock skew or adjustments.

-----

 I’d love to see `(self.performance.now() + self.performance.timeOrigin)` instead of `Date.now()` in my processor code, as my `process` function is called every other third milliseconds at the sample rate of 192000 Hz.

-----

### Note

Currently there’s no way for my processor to use it.

```
DOMException: Failed to execute 'postMessage' on 'MessagePort': Performance object could not be cloned.
```

**Comment by hoch** [comment]
@juj also asked this in #2527. I think we should get the spec change first since the actual spec work might be very small.

**Comment by Night12138** [comment]
Hi everyone, how is the progress going? We need to know the current render capacity, and a high-precision clock is required.

**Comment by hoch** [comment]
2023 TPAC Audio WG Discussion:
Because this involves the high resolution timer on a high priority thread, each implementer will first have an internal security discussion. The discussion will continue once the implementers are confident in changing the spec and beginning implementation.

**Comment by guest271314** [comment]
There are multiple ways this can be implemented in a _user-defined_ manner.  I think the simplest way is a [webmanifest](https://w3c.github.io/manifest/) that Progressive Web Apps and Isolated Web Apps use, e.g., Isolated Web Apps have something like

```
  "permissions_policy": {
    "cross-origin-isolated": ["self"],
    "direct-sockets": ["self"]
  }
```

where for `AudioWorkletGlobalScope` we could do something like

```
"exposed":  {
   "performance": true
   "atob": true,
   "fetch": true
}
```

Done. That's it. The user defines whether a Web API or global function is exposed, or not.

**Comment by padenot** [comment]
> Because this involves the high resolution timer on a high priority thread, each implementer will first have an internal security discussion. The discussion will continue once the implementers are confident in changing the spec and beginning implementation.

With my Firefox implementer hat on, and after talking internally:

In Firefox, iff if the page is able to instantiate `SharedArrayBuffer` (e.g. because the headers have been set when loading the document), then the precision of `performance.now()` is 20us, which might well be good enough for our purpose here.

**Comment by hoch** [comment]
@padenot What would be the next step to make this possible? It looks like a great addition that is reasonably popular.

**Comment by sbc100** [comment]
This issue came up again recently on the emscripten side: https://github.com/emscripten-core/emscripten/issues/22962.

**Comment by mjwilson-google** [comment]
We now have a [Chromium issue](https://crbug.com/392542574) that is asking for this and also said that Microsoft Edge has already gone ahead and implemented the Performance API in worklets.

It seems like there is support for this overall so we should probably spec it out soon.  I'll add it to the agenda for next WG meeting.

**Comment by padenot** [comment]
To follow-up to what @mjwilson-google says, this would need to happen here: https://w3c.github.io/hr-time/#sec-performance

Just a PR that adds `, AudioWorklet`, after in the `Exposed` section.

**Comment by chrisguttandin** [comment]
I tried to check if this is really implemeted in Edge. I ran my test in version 132.0.2957.140 on macOS. This is the code that I used.

```js
class TestProcessor extends AudioWorkletProcessor {
    constructor () {
        super();

        console.log(typeof performance);
    }

    process() {
        console.log(typeof performance);

        return true;
    }
}

console.log(typeof performance);

registerProcessor('test-processor', TestProcessor);
```

It logged `'undefined'` all the time. Maybe I'm testing for the wrong thing but it looks like the Performance API isn't available in Edge either.

**Comment by padenot** [comment]
@chrisguttandin did you try to run this in isolated mode as well?

**Comment by mjwilson-google** [comment]
Even if it's not actually supported in Edge yet we should still do this.  Maybe I misunderstood the issue report.

> this would need to happen here: https://w3c.github.io/hr-time/#sec-performance

So it's actually outside our spec?  I can raise the PR if we're all in agreement.

**Comment by padenot** [comment]
> So it's actually outside our spec? I can raise the PR if we're all in agreement.

Yes, this spec I link defined where their stuff is exposed, but we're the one that can ask for it to be exposed, if that makes sense.

**Comment by chrisguttandin** [comment]
> @chrisguttandin did you try to run this in isolated mode as well?

@padenot I set the COEP/COOP header pair and also made sure that `SharedArrayBuffer` is available.

**Comment by mjwilson-google** [comment]
> we're the one that can ask for it to be exposed, if that makes sense.

Makes sense to me!  I raised https://github.com/w3c/hr-time/pull/169 -- please add supporting comments there if necessary.

**Comment by padenot** [comment]
Firefox implementation: https://bugzilla.mozilla.org/show_bug.cgi?id=2001624 (working, not reviewed / merged)
